ex: dockergc: various fixes

sjenning opened the pull request:
Fixes https://github.com/openshift/origin/issues/17443

xref https://bugzilla.redhat.com/show_bug.cgi?id=1532966

@legionus @dmage

legionus commented on contrib/completions/bash/openshift:
```diff
@@ -429,30 +429,12 @@ _openshift_ex_dockergc()
     flags_completion=()
 
-    flags+=("--dry-run")
-    local_nonpersistent_flags+=("--dry-run")
```
No. Don't do it. This is a common option.

legionus commented on pkg/oc/experimental/dockergc/dockergc.go:
```diff
@@ -178,22 +170,22 @@ func doGarbageCollection(ctx context.Context, client *dockerapi.Client, options
 		}
-		fmt.Printf("removing container %v (size: %v, age: %v)\n", c.ID, c.SizeRw, age)
+		glog.Infof("removing container %v (size: %v, age: %v)", c.ID, c.SizeRw, age)
 		err := client.ContainerRemove(ctx, c.ID, dockertypes.ContainerRemoveOptions{RemoveVolumes: true})
```
Do not make any real changes if the `--dry-run` is used.

legionus commented on pkg/oc/experimental/dockergc/dockergc.go:
```diff
@@ -210,27 +202,27 @@ func doGarbageCollection(ctx context.Context, client *dockerapi.Client, options
 		}
-		fmt.Printf("removing image %v (size: %v, age: %v)\n", i.ID, i.Size, age)
+		glog.Infof("removing image %v (size: %v, age: %v)", i.ID, i.Size, age)
 		_, err := client.ImageRemove(ctx, i.ID, dockertypes.ImageRemoveOptions{PruneChildren: true})
```
Do not make any real changes if the `--dry-run` is used.

sjenning commented:
@legionus ok I added the dry-run flag back.  It runs in a single-pass mode with no effect.

I also noticed that I had *completely* misunderstood how contexts work.  I've added a commit to fix that.

sjenning commented:
/retest

sjenning commented:
/retest

sjenning commented:
/retest

dmage commented on pkg/oc/experimental/dockergc/client.go:
```diff
@@ -0,0 +1,90 @@
+	ctx, cancel := c.getTimeoutContext()
+	defer cancel()
+	containers, err := c.client.ContainerList(ctx, options)
+	if ctx.Err() == context.DeadlineExceeded {
```
Why do you need this additional check?

sjenning replied:
There are two places an error can happen: the ctx can timeout or the client operation an error.  We just need to check both.  AFAICT `DeadlineExceeded` and `Canceled` are the only possible context errors.  Could probably just change this to `ctx.Err() != nil`, but since there is no way for the context to be cancelled here since the call to `cancel()` is a defer for this function, they are functionally equivalent.

Since all tests are green, I'd like to avoid non-functional changes if we can.  This ok?

dmage replied:
If the context is cancelled/timed out, ContainerList should return an error. Though, if it isn't, why do you want to abandon the result in this case?

sjenning replied:
fair enough. updated.

legionus replied:
```go
func clientErr(ctx context.Context, err error) error {
	if ctx.Err() != nil {
		return ctx.Err()
	}
	return err
}
```
@sjenning You move an error handling to this function, it checks `ctx.Err()` first even if `err == nil` in the function arguments. I do not see any changes in the old logic.

legionus replied:
@sjenning ping

sjenning commented:
@legionus look good to you now?

eparis commented:
/approve

legionus commented:
> look good to you now?

LGTM but I think that @dmage's comments should be addressed.

sjenning commented:
@derekwaynecarr can I get lgtm?

sjenning commented:
flake https://github.com/openshift/origin/issues/17698
/retest

sjenning commented:
opened e2e flake https://github.com/openshift/origin/issues/17721
/retest

sjenning commented:
/test end_to_end

sjenning commented:
xref https://bugzilla.redhat.com/show_bug.cgi?id=1532966

eparis commented:
/lgtm

derekwaynecarr commented:
/lgtm
